docs/articles/layouts.md

```diff
@@ -357,7 +359,99 @@ code (please note, that the value of a variable can be an expression):
 Parameterized fragments let developers create functional-like fragments
 that are easier to reuse. Read more about parameterized fragments in the
 Thymeleaf documentation: [Parameterizable fragment
-signatures](http://www.thymeleaf.org/doc/html/Using-Thymeleaf.html#parameterizable-fragment-signatures).
+signatures](http://www.thymeleaf.org/doc/tutorials/3.0/usingthymeleaf.html#parameterizable-fragment-signatures).
+
+### Fragment Expressions
+
+Thymeleaf 3.0 introduced a new type of expression as a part of the general Thymeleaf Standard Expression system: **Fragment Expressions**: 
+
+```xml
+    <div th:insert="~{fragments/footer :: footer}">...</div>
+```
+
+The idea of this syntax is to be able to use resolved fragments as any other kind of objects in the template execution context for later use:
+
+```xml
+<div th:replace="${#authentication.principal.isAdmin()} ? ~{fragments/footer :: footer-admin} : ~{fragments/footer :: footer-admin}">
+  &copy; 2016 The Static Templates
+</div>
+```
+
+Fragment expression allows creating fragments in a way such that they can be enriched with markup coming from the calling templates, resulting in a layout mechanism that is far more flexible than `th:insert` and `th:replace`only.
+
+#### Flexible Layout Example
+
+The `task/layout.html` file defines all the fragments that will be used by calling templates. The below `header` fragment takes `breadcrumb` parameter that will replace `ol` markup with its resolved value:
+
+```xml
+<!--/* Header fragment */-->
+<div th:fragment="header(breadcrumb)">
+    <ol class="breadcrumb container" th:replace="${breadcrumb}">
+        <li><a href="#">Home</a></li>
+    </ol>
+</div>
+```
+In the calling template (`task/task-list.html`) we will use a *Markup Selector* syntax to pass the element matching `.breadcrumb` selector:
+
+```xml
+<!--/* The markup with breadcrumb class will be passed to the header fragment */-->
+<header th:insert="task/layout :: header(~{ :: .breadcrumb})">
+    <ol class="breadcrumb container">
+        <li><a href="#">Home</a></li>
+        <li><a href="#" th:href="@{/task}">Tasks</a></li>
+    </ol>
+</header>
+```
+
+As a result, the following HTML will be generated for the `task/taks-list` view:
+
+```xml
+<header>
+    <ol class="breadcrumb container">
+        <li><a href="#">Home</a></li>
+        <li><a href="[...]">Tasks</a></li>
+    </ol>
+</header>
+```
```

Comment:
Oh wait, I just did, silly me. Okay, I can explain this, it's actually expected behaviour.

The fact is that this:
```html
<header th:insert="task/layout :: header (~{ :: .breadcrumb})">
```
...is an expression that contains an *unnamed* parameter. Which is fine because `header` is the name of a fragment defined with a `th:fragment`, and in such `th:fragment` there already is a name for such fragment (`breadcrumb`):
```html
<div th:fragment="header(breadcrumb)">
```
This makes this `th:replace` make sense:
```html
<ol class="breadcrumb container" th:replace="${breadcrumb}">
```

**BUT**, when the `th:insert` expression is this one:
```html
<header th:insert="task/layout :: header/content() (~{ :: .breadcrumb})">
```
That still has an *unnamed* parameter, but the `header/content()` selector does not select any `th:fragment`, but instead the contents of one. A collection of nodes.

And such collection of nodes does not have a *signature*. So there is no way to map a *parameter name* to the unnamed parameter specified in the `th:insert`, which is why this fails:
```html
<ol class="breadcrumb container" th:replace="${breadcrumb}">
```
Note it is the `th:replace` that is failing, because it doesn't find a template/fragment with name `""`, because `breadcrumb` has no value. Because we haven't got *through the signature*. Because we haven't selected it, but merely used it as a markup reference.

Easy solution? simply apply a name to that parameter so that signature is unneeded (actually, you can remove the parameter name from the fragment):
```html
<header th:insert="task/layout :: header/content() (breadcrumb=~{ :: .breadcrumb})">
```
...and then no parameters needed here (wouldn't actually do anything):
```html
<div th:fragment="header">
```

This anyway makes me think that maybe this should be warned to the user, so that an exception is thrown when a fragment *without a signature* is called with unnamed parameters. Strangely, I remember that code existing at some point in the past… maybe it existed in 2.1 and I removed it for some reason? I will need to check.